sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -396,6 +398,44 @@ public BoundedWindow window() {
 
 
   /**
+   * Annotation for declaring and dereferencing state cells. <i>Not currently supported by any
+   * runner</i>.
+   *
+   * <p>To declare a state cell, create a field of type a {@link StateSpec} and annotate it
+   * with a {@link StateId} like so:
+   *
+   * <code>
+   *   new DoFn<KV<Key, Foo>, Baz>() {
+   *     @StateId("my-state-id")
+   *     private final StateSpec<K, ValueState<MyState>> myStateSpec =
+   *         StateSpecs.value(new MyStateCoder());
+   *   }
+   * </code>
+   *
+   * <p>To use a declared state cell during processing, add a parameter of the appropriate
```

Comment:
May be easier as a single state example?

```
To use state within a {@link DoFn}, create a field of type {@link StateSpec} and annotate it with a {@link StateId}. To access the state within a {@link ProcessElement @ProcessElement} call add a parameter of the appropriate {@link State} subclass and annotate it with the {@link StateId}:

...
```

May want to document some of the future restrictions:
- Having a StateSpec field without a StateId is an error.
- Having a StateSpec field that isn't used in any of the processing functions is an error.
- Having multiple fields with the same StateId is an error.
- Having the StateSpec field be non-final is an error. 
- Accessing a StateId that doesn't have a corresponding field is an error.
- Having a parameter that has a different type of state than the associated field is an error.
...